File: LATPRAK4DPBO/C++/ParkingLot.cpp
// Saya Marvel Ravindra Dioputra [2200481] LatPrak4 C++ dalam Mata Kuliah DPBO 
// untuk keberkahanNya maka saya tidak melakukan kecurangan seperti yang telah dispesifikasikan. Aamiin.

#pragma once
#include "Vehicle.cpp"  // composite bahwa parking lot mempunyai vehicle
#include <vector>       // untuk list menampung kendaraan pada parking lot

class ParkingLot {
private:
    int kapasitas;                      // atribut kapasitas parking lot
    vector<Vehicle> daftarKendaraan;    // atribut list kendaraan

public:
    // konstruktor
    ParkingLot() {

    }

    ParkingLot(int kapasitas) : kapasitas(kapasitas) {}

    // method untuk menambahkan kendaraan ke parkiran
    void tambahKendaraan(Vehicle kendaraan) {
        if (daftarKendaraan.size() >= kapasitas) {
            cout << "Parking Lot penuh" << endl;
        } else {
            daftarKendaraan.push_back(kendaraan);
            cout << "Kendaraan masuk ke Parking Lot. Kaspasitas: " << kapasitas << ", Jumlah Kendaraan: " << daftarKendaraan.size() << endl;
        }
    }

    // method untuk menampilkan kendaraan yang ada pada parkiran tersebut
    void tampilkanKendaraan() {
        cout << "Daftar Kendaraan di Parking Lot:" << endl;
        for (int i = 0; i < daftarKendaraan.size(); i++) {
            cout << "Plat Nomor: " << daftarKendaraan[i].getPlat() << ", Merk: " << daftarKendaraan[i].getMerk() << ", Tahun Produksi: " << daftarKendaraan[i].getTahunProduksi() << ", Warna: " << daftarKendaraan[i].getWarna() << endl;
        }
    }

    // method untuk mengeluarkan kendaraan dari parkiran berdasarkan plat nomor yang di cari
    void keluarParkir(string platNomor) {
        for (int i = 0; i < daftarKendaraan.size(); i++) {
            if (daftarKendaraan[i].getPlat() == platNomor) {
                daftarKendaraan.erase(daftarKendaraan.begin() + i);
                cout << "Kendaraan dengan plat nomor " << platNomor << " telah keluar dari Parking Lot" << endl;
                cout << "Kapasitas: " << kapasitas << ", Jumlah Kendaraan: " << daftarKendaraan.size() << endl;
                return;
            }
        }
        cout << "Kendaraan dengan plat nomor " << platNomor << " tidak ditemukan di Parking Lot" << endl;
    }

    // destruktor
    ~ParkingLot() {
        
    }
};

File: LATPRAK4DPBO/python/parking_lot.py
# Saya Marvel Ravindra Dioputra [2200481] LatPrak4 Python dalam Mata Kuliah DPBO 
# untuk keberkahanNya maka saya tidak melakukan kecurangan seperti yang telah dispesifikasikan. Aamiin.

# class parking lot
class ParkingLot:
    # inisialisasi/instansiasi
    def __init__(self, kapasitas):
        self.kapasitas = kapasitas
        self.jumlah_kendaraan = 0

    # method untuk menambahkan kendaraan ke parkiran parking lot
    def park_vehicle(self, vehicle):
        # jika masih kosong
        if self.jumlah_kendaraan < self.kapasitas:
            self.jumlah_kendaraan += 1
            print(f"{vehicle.__class__.__name__} dengan plat nomor {vehicle.plat_nomor} berhasil diparkir.")
        # jika sudah penuh
        else:
            print("Parking lot penuh")

    # method untuk kendaraan keluar parkiran
    def leave_parking(self, vehicle):
        if self.jumlah_kendaraan > 0:
            self.jumlah_kendaraan -= 1
            print(f"{vehicle.__class__.__name__} dengan plat nomor {vehicle.plat_nomor} meninggalkan parking lot.")
        else:
            print("Parking lot kosong.")

    # method menampilkan data keterangan parking lot
    def display_availability(self):
        print("Kapasitas:", self.kapasitas)
        print("Jumlah Kendaraan:", self.jumlah_kendaraan)
        print("Tersisa:", self.kapasitas - self.jumlah_kendaraan)   
